Makefile

```diff
@@ -71,5 +77,11 @@ curl_local:
 		"W":["DA", "S4", "HT", "C5", "D4", "D7", "S6", "S3", "DK", "CT", "D2", "SK", "H8"],    \
 		"N":["C7", "H6", "H7", "H9", "CJ", "SA", "S8", "SQ", "D5", "S5", "HK", "C8", "HA"],    \
 		"E":["H2", "H5", "CQ", "D9", "H4", "ST", "HQ", "SJ", "HJ", "DQ", "H3", "C9", "CK"] }}' \
-	http://localhost:5000/api/dds-table/
+	${URL}
+
+curl_local:
+	make curl URL=${LOCAL_URL}
```

Comment:
Recursive make is wrong in most cases. It should be avoided pretty much always.

Better alternative would be something like
```
CURL_CMD := curl --header ...

curl_local:
    ${CURL_CMD} ${LOCAL_URL}

Reply:
I like your suggestion but it has a problem. When I use the recursive make I get a nicely formatted command line:

```
Ragnar:DDS-aw adamw$ make curl_prod
make curl URL=https://dds.prod.globalbridge.app/api/dds-table/
curl \
	--header "Content-Type: application/json" \
	--data \
	'{"hands": { \
		"S":["D3", "C6", "DT", "D8", "DJ", "D6", "CA", "C3", "S2", "C2", "C4", "S9", "S7"],    \
		"W":["DA", "S4", "HT", "C5", "D4", "D7", "S6", "S3", "DK", "CT", "D2", "SK", "H8"],    \
		"N":["C7", "H6", "H7", "H9", "CJ", "SA", "S8", "SQ", "D5", "S5", "HK", "C8", "HA"],    \
		"E":["H2", "H5", "CQ", "D9", "H4", "ST", "HQ", "SJ", "HJ", "DQ", "H3", "C9", "CK"] }}' \
	https://dds.prod.globalbridge.app/api/dds-table/
```

When I use either your suggestion or my earlier attempt with a DEAL macro, I get this:

```
Ragnar:DDS-aw adamw$ make curl_prod
curl --header "Content-Type: application/json" --data '{"hands": { "S":["D3", "C6", "DT", "D8", "DJ", "D6", "CA", "C3", "S2", "C2", "C4", "S9", "S7"], "W":["DA", "S4", "HT", "C5", "D4", "D7", "S6", "S3", "DK", "CT", "D2", "SK", "H8"], "N":["C7", "H6", "H7", "H9", "CJ", "SA", "S8", "SQ", "D5", "S5", "HK", "C8", "HA"], "E":["H2", "H5", "CQ", "D9", "H4", "ST", "HQ", "SJ", "HJ", "DQ", "H3", "C9", "CK"] }}'  https://dds.prod.globalbridge.app/api/dds-table/
```

Switching from **:=** to **=** did not make a difference.

Do you know a good way around this? 

I found Peter Miller's [Recursive Make Considered Harmful](http://www.real-linux.org.uk/recursivemake.pdf) article and I don't think his argument applies to a case like this.

Reply:
> I like your suggestion but it has a problem. When I use the recursive make I get a nicely formatted command line:
> 
> Switching from **:=** to **=** did not make a difference.
> 

You could split the command to multiple variables. Then use backslash after each to get multiline output.

> Do you know a good way around this?
> 
> I found Peter Miller's [Recursive Make Considered Harmful](http://www.real-linux.org.uk/recursivemake.pdf) article and I don't think his argument applies to a case like this.

I don't remember what was there. Example for relevant issues is that cmake Makefile backend can end up running same commands multiple times because it use very similar approach of recursing to same makefile multiple times. (One reason why I only started liking CMake after they added Ninja backend)

Reply:
Something like this?

```
curl_local:
	${CURL_CMD} \
	'{"hands": {     \
		${SOUTH},    \
		${WEST},     \
		${NORTH},    \
		${EAST}  }}' \
	${LOCAL_URL}
```

It works, but the recursive make call is shorter and seems cleaner.

Reply:
(These targets will not interact with CMake.)

Reply:
I've asked for suggestions:

https://stackoverflow.com/questions/62827211/how-to-embed-newlines-in-a-macro-inside-a-recipe

Reply:
> (These targets will not interact with CMake.)

True. But same risk are involved here as with recursion generated by CMake.

> I've asked for suggestions:
> 
> https://stackoverflow.com/questions/62827211/how-to-embed-newlines-in-a-macro-inside-a-recipe

I think the better solution to clean up makefile is the silent make the target idiom. http://make.mad-scientist.net/managing-recipe-echoing/

If makefile requires many calls to same tool then it could includes something like:
``
ifeq (V,)
SCURL = @echo CURL $@ && curl
else
SCURL = curl
endif

CURL_CMD = ${SCURL} --header ....
```
Does that default clean output with `make <targets> V=1` as a potential easy full command output for debugging?

Reply:
> I've asked for suggestions:
> 
> https://stackoverflow.com/questions/62827211/how-to-embed-newlines-in-a-macro-inside-a-recipe

Oops. My Makefile knowledge is fading. I forgot you can do something like
```
define CURL_CMD
curl --header ... \
 '{"hands": { ...\
... }'
$1
enddef

local_curl:
    $(call CURL_CMD,${URL})
```

Reply:
Neither of these gives the output I'm hoping for. _Make_ seems to turn newlines to spaces any time it expands a macro in a recipe.

The first StackExchange answer suggests:

_But to be honest, I would try to explain to your reviewer that fiddling with technically superfluous spaces and newlines in makefiles is lowering the quality of a makefile instead of raising it._

The recursive make solution is the only one where I've been able to get clean output, and I don't think the general objections apply since this command is simply "executable documentation" and does not involve any dependency analysis. Conceptually it could just as easily be accomplished with shell commands, but that would clutter our directory.

Reply:
> The recursive make solution is the only one where I've been able to get clean output, and I don't think the general objections apply since this command is simply "executable documentation" and does not involve any dependency analysis. Conceptually it could just as easily be accomplished with shell commands, but that would clutter our directory.

Makefiles either declare dependencies or become quickly hard to use. That is my experience with all Makefiles.

What do you say about following output?

```
CURL   http://localhost:5000/api/dds-table/
curl: (7) Failed to connect to localhost port 5000: Connection refused
make: *** [Makefile:88: curl_local] Error 7
```

It is produced from a silent makefile.

```
ifeq (V,1)
SCURL = curl
else
SCURL = @echo "CURL   ${1}" && curl
endif

define CURL_CMD
	${SCURL} \
	--header "Content-Type: application/json" \
	--data \
	'{"hands": { \
		"S":["D3", "C6", "DT", "D8", "DJ", "D6", "CA", "C3", "S2", "C2", "C4", "S9", "S7"], \
		"W":["DA", "S4", "HT", "C5", "D4", "D7", "S6", "S3", "DK", "CT", "D2", "SK", "H8"], \
		"N":["C7", "H6", "H7", "H9", "CJ", "SA", "S8", "SQ", "D5", "S5", "HK", "C8", "HA"], \
		"E":["H2", "H5", "CQ", "D9", "H4", "ST", "HQ", "SJ", "HJ", "DQ", "H3", "C9", "CK"] }}' \
	${1}
endef

curl_local:
	$(call CURL_CMD,${LOCAL_URL})

curl_prod:
	$(call CURL_CMD,${PROD_URL})
```

Reply:
This is the command output I'm hoping for. It shows the user exactly what _make_ is executing on his behalf.

```
Ragnar:DDS-aw adamw$ make curl_prod
make curl URL=https://dds.prod.globalbridge.app/api/dds-table/
curl \
	--header "Content-Type: application/json" \
	--data \
	'{"hands": { \
		"S":["D3", "C6", "DT", "D8", "DJ", "D6", "CA", "C3", "S2", "C2", "C4", "S9", "S7"],    \
		"W":["DA", "S4", "HT", "C5", "D4", "D7", "S6", "S3", "DK", "CT", "D2", "SK", "H8"],    \
		"N":["C7", "H6", "H7", "H9", "CJ", "SA", "S8", "SQ", "D5", "S5", "HK", "C8", "HA"],    \
		"E":["H2", "H5", "CQ", "D9", "H4", "ST", "HQ", "SJ", "HJ", "DQ", "H3", "C9", "CK"] }}' \
	https://dds.prod.globalbridge.app/api/dds-table/
```

How about I submit the recursive version and agree to change it if it ever becomes a problem?

Reply:
Actually, the first line of output is unnecessary. It's the curl command itself I want to show.

Reply:
Then there is non silent correctly formatted version without duplication. I forgot how to use make variables effectively.

```
curl_local_URL := http://localhost:5000/api/dds-table/
curl_prod_URL  := https://dds.prod.globalbridge.app/api/dds-table/
CURL_URL ?= ${$@_URL}

# We'd like to define a TEST_DEAL macro, but then the command-line output is
# not as pretty.
curl_local curl_prod:
	curl \
	--header "Content-Type: application/json" \
	--data \
	'{"hands": { \
		"S":["D3", "C6", "DT", "D8", "DJ", "D6", "CA", "C3", "S2", "C2", "C4", "S9", "S7"],    \
		"W":["DA", "S4", "HT", "C5", "D4", "D7", "S6", "S3", "DK", "CT", "D2", "SK", "H8"],    \
		"N":["C7", "H6", "H7", "H9", "CJ", "SA", "S8", "SQ", "D5", "S5", "HK", "C8", "HA"],    \
		"E":["H2", "H5", "CQ", "D9", "H4", "ST", "HQ", "SJ", "HJ", "DQ", "H3", "C9", "CK"] }}' \
	${CURL_URL}
```
`$@` gets expanded first to target and then the correct URL variable is expanded.

Reply:
That looks perfect! I'll try it and report back.

Reply:
Interesting discussion, I'm at this step too. This PR works for me so far building and testing locally.

Reply:
Works nicely - thanks for persevering. I'll rebase later today.

Reply:
This is working for me

CURL_ARGS ?= --header "Content-Type: application/json" \
    --data \
    '{"hands": { \
        "S":["D3", "C6", "DT", "D8", "DJ", "D6", "CA", "C3", "S2", "C2", "C4", "S9", "S7"],    \
        "W":["DA", "S4", "HT", "C5", "D4", "D7", "S6", "S3", "DK", "CT", "D2", "SK", "H8"],    \
        "N":["C7", "H6", "H7", "H9", "CJ", "SA", "S8", "SQ", "D5", "S5", "HK", "C8", "HA"],    \
        "E":["H2", "H5", "CQ", "D9", "H4", "ST", "HQ", "SJ", "HJ", "DQ", "H3", "C9", "CK"] }}'

curl:
    curl ${CURL_ARGS} ${URL}

Reply:
There is a minor difference:
```
$ make test1 test2 -f test.Makefile 
echo foo bar
echo foo \
	bar
foo bar
foo bar
$ cat test.Makefile 
ARGS ?= foo \
		bar

test1:
	echo ${ARGS}

test2:
	echo foo \
		bar
```
Adam wants to see the long command line split to multiple lines in the output. I agree with his wish. Figuring out the simplest solution was surprising hard.